### DISCUSSION TITLE
Which type of component should I use to display dynamic content  on the homepage?

### DISCUSSION THREAD

**Discussion by sipronunciaaigor** [OrchardCMS/OrchardCore#11909] [Q&A] [answered]
Hi, I am trying to enrich the homepage of my orchard core project with some dynamic areas.
I applied the Blog theme as a starting point and I customised my inner pages and navigation and I am using a mix of razor and liquid (it's be a bit messy, but it works so far).
The website is almost entirely data driven from a source of data which is entirely separated from orchard core.
My problems arise when I want to start mixing on the homepage some areas dynamically provided by orchard core and some of my data-driven bits (rest of the website is ok as it's well decoupled) and that's where I found myself in need to ask for some kind help from the community.

The following image is a sort of shallow wireframe of how I want my homepage to behave.

![wireframe](https://user-images.githubusercontent.com/16055024/176034694-fb5e7806-ad79-4493-a89a-a06577384d36.jpg)

I have problems with the blue areas I highlighted as "content" followed by a route.
The idea is that I want "some sort of component" in the admin area where I can set the route value and, as an effect, I want its counterpart on the website homepage to display the content of that route that has been set in the admin.
Furthermore I want to be able to add/remove as many of this kind of components as I want.

All routes will return the same shape of results as they rely on services that read data from non-orchard areas and, for that purpose, they all return a `List<MyItem>`.
The content might change by the minute (hence the route based on underlying services).
The goal is that, on the homepage, they will simply display ordered data from different periods or different aggregation approaches. and they will all have media url, title, short description and a link.
In the very end is essentially a list of links: think of it as if you want to display top scores in the past seven days in one widget, top 5 mobile phones in 2021 in the second one, last 5 wikipedia articles in the third one etc.

I think I should use some sort of custom widget, but, if that's the case, I probably don't understand how they work.

I have several questions here:
1. What is the "component" I should use? All the terms I could think of to describe those areas in the page are already used in the documentation from component, to widget, item, part etc. and I think I got lost/confused as they all have very different meanings one from the other.
2. How do I find (in razor or liquid) the list of these "components", once I'm building my homepage?
3. How do I ensure these "components" are shown only on my homepage and not when I go to the next page of my list of posts? (I guess I'll be able to tick the homepage only checkbox)
4. If the components end up NOT being widgets, how can I reuse them in other areas of the website if needed?

Other discussions or fundamental resources I tried to look at so far:
Documentation (got a bit confused as I didn't understand what/how to use).
Orchard DOJO tutorial
https://github.com/OrchardCMS/OrchardCore/discussions/9846
https://github.com/OrchardCMS/OrchardCore/discussions/11225

**Comment by Skrypt** [comment, marked as answer]
You can retrieve a list of content items by using the SQL or Lucene Queries and the Query API. These are documented.

**Reply by sipronunciaaigor** [reply]
Sorry @Skrypt, but I fear I don't think I get it exactly...
My data can't be retrieved from Lucene or any Orchard tables as my content is serviced from other sources. Think of it as if you are preparing a reporting system for a company. Company data wouldn't be part of the CMS, but it would likely come from sever endpoints from several microservices.
Or you are referring just to my second question: `How do I find (in razor or liquid) the list of these "components", once I'm building my homepage?`? In that case I guess I can refer to https://www.youtube.com/watch?v=JYES1i6BdWs after I create a custom content item... (which leads me to think I might need to use a content item as one of the "things" I need to achieve my goal...)

**Reply by Skrypt** [reply]
The Lucene and SQL Queries always return an IQueryResults. It is not given that these results will be parsed from JSON to a Content Item. But the CMS features are generally making use of Content Items that are defined by the CMS and the data is generally parsed to a Content Item in that case. So, if you want to work with data that comes from different sources then you may need to simply create a new DataProvider that returns IQueryResults. Currently, the SQL Query feature only supports using the current tenant connection string. Maybe, having the option to use multiple database connection strings would be an option to add. Though, if you are using microservices then the DAL of these should be abstracted and taken care of. All you need is to use them by calling their public methods and displaying their data in a custom Orchard Core module. Now, if you want to display this data on your Home Page through the use of a Widget you can create a custom Widget that displays this data. You can create a custom Shape that will render this data or a TagHelper for Liquid and Razor views to render this component.

**Reply by sipronunciaaigor** [reply]
Amazing! Thanks heaps for the reply.
I didn't about the possibility to swap the DataProvider. I'll give it a go.
Thanks also for confirming the type of component that I need to focus on: the widget.
I think I missed the custom shape that needs to wrap the widget.
On the other hand, I think never touched any TagHelper, so I will, as well, try to dig in that direction.
Thanks again.

p.s.: I'll mark the initial reply as the answer to highlight this innermost reply
